### ISSUE TITLE
Inconsistent array indexing in -nonfScalarize flat output

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#8128]
Consider the following model:
```Modelica
model M
  model Q
    Real x;
  equation
    x = 1;
  end Q;
  model P
    Real x[2];
  equation
    for i in 1:2 loop 
      x[i] = 2;
    end for;
  end P;
  Q q[2];
  P p;
end M;
```
The flat output with `-nonfScalarize` is
```Modelica
class M
  Real[2] q.x;
  Real[2] p.x;
equation
  for $i1 in 1:2 loop
    q[$i1].x = 1.0;
  end for;
  for i in 1:2 loop
    p.x[i] = 2.0;
  end for;
end M;
```
The declarations of `q.x` and `p.x` are identical: they are two arrays of two Reals. This is perfectly fine, since from a mathematical point of view (which is what matters to run a simulation), that's what they are. However, those variables are rendered differently in the equation section. This is no big deal if the file is to be read by a human, but if we want a compiler to read them, it's a bit awkward. 

Couldn't we get
```Modelica
class M
  Real[2] q.x;
  Real[2] p.x;
equation
  for $i1 in 1:2 loop
    q.x[$i1] = 1.0;
  end for;
  for i in 1:2 loop
    p.x[i] = 2.0;
  end for;
end M;
```
instead?

Also, I understand the index `$i1` starts with a dollar as all other identifiers that are not originally present in the model, such as `$cse1`. Can you please confirm that?

Thanks!

**Comment by perost** [comment]
You can use `-d=combineSubscripts` to move all the subscripts to the end.

And yes, we typically add `$` to compiler-generated identifiers since it's not valid Modelica and thus doesn't risk name collisions with existing identifiers.

**Comment by casella** [comment]
> You can use `-d=combineSubscripts` to move all the subscripts to the end.

Aha, I was not aware of that. Or maybe I was, and then forgot about it.

> And yes, we typically add `$` to compiler-generated identifiers since it's not valid Modelica and thus doesn't risk name collisions with existing identifiers.

OK.

**Comment by perost** [comment]
> > You can use `-d=combineSubscripts` to move all the subscripts to the end.
> 
> Aha, I was not aware of that. Or maybe I was, and then forgot about it.

Well, I implemented it for you in [#6173](https://trac.openmodelica.org/OpenModelica/ticket/6173) :smile: That ticket is still open btw.

**Comment by casella** [comment]
> Well, I implemented it for you in [#6173](https://trac.openmodelica.org/OpenModelica/ticket/6173) 😄 

In fact, I had a sense of deja-vù here. I'm getting old...

> That ticket is still open btw.

OK, I'll comment there. There are still some problems with subscripts when also using -f and -d=arrayConnect, see #8130.
